src/test/groovy/io/curity/identityserver/plugin/OptInMFA/OptInMFAAuthenticationActionTest.groovy

```diff
@@ -0,0 +1,242 @@
+/*
+ *  Copyright 2020 Curity AB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.curity.identityserver.plugin.OptInMFA
+
+import se.curity.identityserver.sdk.attribute.AccountAttributes
+import se.curity.identityserver.sdk.attribute.Attribute
+import se.curity.identityserver.sdk.attribute.AuthenticationAttributes
+import se.curity.identityserver.sdk.attribute.ContextAttributes
+import se.curity.identityserver.sdk.authentication.AuthenticatedSessions
+import se.curity.identityserver.sdk.authenticationaction.AuthenticationActionResult
+import se.curity.identityserver.sdk.authenticationaction.completions.RequiredActionCompletion
+import se.curity.identityserver.sdk.errors.AuthenticatorNotConfiguredException
+import se.curity.identityserver.sdk.service.AccountManager
+import se.curity.identityserver.sdk.service.SessionManager
+import se.curity.identityserver.sdk.service.authenticationaction.AuthenticatorDescriptor
+import se.curity.identityserver.sdk.service.authenticationaction.AuthenticatorDescriptorFactory
+import spock.lang.Shared
+import spock.lang.Specification
+
+import static io.curity.identityserver.plugin.OptInMFA.OptInMFAAuthenticationAction.IS_SECOND_FACTOR_CHOSEN_ATTRIBUTE
+import static io.curity.identityserver.plugin.OptInMFA.OptInMFAAuthenticationAction.CHOSEN_SECOND_FACTOR_ATTRIBUTE
+
+final class OptInMFAAuthenticationActionTest extends Specification {
+
+    private def username = "john"
+    private def authenticationAttributes = AuthenticationAttributes.of(username, ContextAttributes.empty())
+    @Shared private def authenticator = Stub(AuthenticatorDescriptor)
+
+    def setupSpec()
+    {
+        authenticator.getAcr() >> "email"
+    }
+
+    def "should fail authentication when user does not have secondary factors"()
+    {
+        // TODO: this should eventually redirect to registration
+        given:
+        def sessionManager = getSessionManagerStubWithoutChosenSecondFactor()
+
+        def user = getUserAttributes()
+        def accountManager = getAccountManagerStubReturningUser(user)
+
+        def configuration = new TestActionConfiguration(accountManager, null, sessionManager)
+        def action = new OptInMFAAuthenticationAction(configuration)
+
+        when:
+        def result = action.apply(authenticationAttributes, null, "transactionId", null)
+
+        then:
+        result instanceof AuthenticationActionResult.FailedAuthenticationActionResult
```

Comment:
Why does this fail?

Reply:
For now it fails, as there are no means to register new second factors on a user account. There's a TODO at the beginning of this test with a note that eventually this behaviour should be removed.